### PR TITLE
Refactor Push and Pull CLI commands

### LINKED ISSUE #3532: Refactor Pull and Push CLI Code
As brought up in #3526:
Currently the code in `/cmd/internal/cli/pull.go` and `/cmd/internal/cli/push.go` contains very long switch case statements that make it hard to follow. These individual case statements should be broken into functions in order to make them more readable and testable. These functions should live in `/internal/app/singularity`.

### PR DESCRIPTION
**Pull request by ikaneshiro**
**Description of the Pull Request (PR):**

- Takes long cases from switch statements in cli for `push` and `pull` commands and moves them into `internal/app/singularity` to make them easier to read and test in the future.

**This fixes or addresses the following GitHub issues:**

- Fixes #3532 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: cmd/internal/cli/pull.go
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Does this need a message? Like:

```suggestion
			sylog.Fatalf("Unable to pull contianer: %v", err)
```

File: cmd/internal/cli/pull.go
Comment (WestleyK) [inline cmd/internal/cli/pull.go:230]: If go will let you...

```suggestion
		err := singularity.OrasPull(name, ref, force, ociAuth)
```

File: cmd/internal/cli/pull.go
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Maybe here too?

```suggestion
			sylog.Fatalf("Unable to pull oci image: %v", err)
```

File: cmd/internal/cli/push.go
Comment (WestleyK) [inline cmd/internal/cli/push.go]:
```suggestion
				sylog.Fatalf("Unable to push image: %v", err)
```

File: cmd/internal/cli/push.go
Comment (WestleyK) [inline cmd/internal/cli/push.go]: Here too?

```suggestion
				sylog.Fatalf("Unable to push oras image: %v", err)
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
			return fmt.Errorf("cached file hash(%s) and expected hash(%s) does not match", cacheFileHash, libraryImage.Hash)
￼
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
			return fmt.Errorf("error getting imageHash: %v", err)
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
					return fmt.Errorf("unable to delete the container: %v", err)
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return fmt.Errorf("unable to parse oci reference: %s", err)
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return fmt.Errorf("failed to get working directory: %s", err)
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
			return fmt.Errorf("image file already exists - will not overwrite")
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return fmt.Errorf("unable to create new build: %v", err)
```

**Review by WestleyK — CHANGES_REQUESTED** [review]
Couple small suggestions, otherwise looks good.

File: cmd/internal/cli/pull.go
Comment (ikaneshiro) [reply cmd/internal/cli/pull.go:230]: `err` already exists in this scope, I could embed these calls in an `if` statement to limit the scope, but that would mean some pretty long lines

File: internal/app/singularity/pull.go
Comment (mem) [inline internal/app/singularity/pull.go]: The comment should refer to the unusual choice of 0777 for what should be a regular file, noting that the SIF file is created with a shebang to be able to run the image as a "script". I believe this comment was the in the original version.

File: internal/app/singularity/pull.go
Comment (mem) [inline internal/app/singularity/pull.go]: again, I believe this was here before, but the mode argument does not apply in this case.

**Review by mem — COMMENTED** [review]
LGTM, I'll let @WestleyK decide

File: internal/app/singularity/pull.go
Comment (gvallee) [inline internal/app/singularity/pull.go]:
```suggestion
			resp, err := sypgp.AskQuestion("Do you want to proceed? [N/y] ")
```

File: internal/app/singularity/pull.go
Comment (gvallee) [inline internal/app/singularity/pull.go]:
```suggestion
			// if there is a warning, return set error to indicate exit 1
```

File: internal/app/singularity/pull.go
Comment (gvallee) [inline internal/app/singularity/pull.go]: Would it be possible to improve the comment? I personally do not understand what it really means.

File: internal/app/singularity/push.go
Comment (gvallee) [inline internal/app/singularity/push.go]:
```suggestion
// OrasPush uploads the image specified by file and pushes it to the provided oci reference,
```

**Review by gvallee — DISMISSED** [review]
I only did a quick read-through and marked a few quick points to fix. I can check the semantics if required, just let me know.

File: cmd/internal/cli/pull.go
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Does this need:

```suggestion
 			sylog.Fatalf("Unable to download shub image: %v", err)
```

File: cmd/internal/cli/pull.go
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Here too?

```suggestion
 			sylog.Fatalf("Unable to ...: %v\n", err)
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Can you just return the err? Instead of passing it thrught `fmt.Errorf()`?

```suggestion
		return err
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Maybe add:

```suggestion
			return fmt.Errorf("image file already exists: %q - will not overwrite", name)
```

**Review by WestleyK — CHANGES_REQUESTED** [review]
Couple more suggestions :+1:

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return err
```
Or:

```suggestion
		return fmt.Errorf("unable to open file: %v", err)
```

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Same here.

File: internal/app/singularity/pull.go
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Same here.

**Review by WestleyK — CHANGES_REQUESTED** [review]
Couple more suggestions...

**Review by WestleyK — APPROVED** [review]
Looks good to me :+1:

**Comment by ikaneshiro** [comment]
@WestleyK @gvallee going to wait to merge this until #3578 is merged in order to work the conflicts into this refactoring

**Comment by ikaneshiro** [comment]
@WestleyK the merge required a lot of manual conflict resolution, can you give this a quick look to see if I messed anything up?

**Review by WestleyK — APPROVED** [review]
Yup, still looks good to me :+1:

**Pull request merged**
